debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/transforms/type/deserializer/SetTypeDeserializer.java

```diff
@@ -25,29 +25,47 @@
     @Override
     public Object deserialize(AbstractType<?> abstractType, ByteBuffer bb) {
         Set<?> deserializedSet = (Set<?>) super.deserialize(abstractType, bb);
-        List<?> deserializedList = (new ArrayList<>(deserializedSet));
+        List<?> deserializedList = convertDeserializedElementsIfNecessary(abstractType, deserializedSet);
         return Values.convertToList(getSchemaBuilder(abstractType).build(), deserializedList);
     }
 
     @Override
     public SchemaBuilder getSchemaBuilder(AbstractType<?> abstractType) {
-        SetType<?> listType = (SetType<?>) abstractType;
-        AbstractType<?> elementsType = listType.getElementsType();
+        SetType<?> setType = (SetType<?>) abstractType;
+        AbstractType<?> elementsType = setType.getElementsType();
         Schema innerSchema = CassandraTypeDeserializer.getSchemaBuilder(elementsType).build();
         return SchemaBuilder.array(innerSchema).optional();
     }
 
     @Override
-    public Object deserialize(SetType<?> collectionType, ComplexColumnData ccd) {
-        List<ByteBuffer> bbList = collectionType.serializedValues(ccd.iterator());
-        AbstractType<?> innerType = collectionType.getElementsType();
-
+    public Object deserialize(SetType<?> setType, ComplexColumnData ccd) {
+        List<ByteBuffer> bbList = setType.serializedValues(ccd.iterator());
+        AbstractType<?> elementsType = setType.getElementsType();
         Set<Object> deserializedSet = new HashSet<>();
         for (ByteBuffer bb : bbList) {
-            deserializedSet.add(super.deserialize(innerType, bb));
+            deserializedSet.add(CassandraTypeDeserializer.deserialize(elementsType, bb));
         }
-
         List<Object> deserializedList = new ArrayList<>(deserializedSet);
-        return Values.convertToList(getSchemaBuilder(collectionType).build(), deserializedList);
+        return Values.convertToList(getSchemaBuilder(setType).build(), deserializedList);
+    }
+
+    /**
+     * If elements in a deserialized set is LogicalType, convert each element to fit in Kafka Schema type
+     * @param abstractType the {@link AbstractType} of a column in Cassandra
+     * @param deserializedSet Set deserialized from Cassandra
+     * @return A deserialized list from Cassandra with each element that fits in Kafka Schema type
+     */
+    private List<Object> convertDeserializedElementsIfNecessary(AbstractType<?> abstractType, Set<?> deserializedSet) {
```

Comment:
How comes a set goes in but a list goes out? Would it make sense to emit sets?

Reply:
> How comes a set goes in but a list goes out? Would it make sense to emit sets?

We don't have a convertToSet method in Kafka Connect library, there's only convertToList, so we have to convert a set to a list in kafka here.